crates/pathfinder/src/state/sync.rs

```diff
@@ -619,6 +728,103 @@ fn deploy_contract(
         .context("Inserting class hash into contracts table")
 }
 
+/// Downloads and inserts class definitions for any classes in the
+/// list which are not already present in the database.
+async fn download_verify_and_insert_missing_classes<
+    SequencerClient: sequencer::ClientApi,
+    ClassIter: Iterator<Item = ClassHash>,
+>(
+    sequencer: SequencerClient,
+    connection: &mut Connection,
+    classes: ClassIter,
+) -> anyhow::Result<()> {
+    use crate::state::class_hash::extract_abi_code_hash;
+
+    // Make list unique.
+    let classes = classes
+        .collect::<std::collections::HashSet<_>>()
+        .into_iter() // TODO: remove this allocation by using Iter in exists.
+        .collect::<Vec<_>>();
+
+    // Check database to see which are missing.
+    let exists = tokio::task::block_in_place(|| {
+        let transaction = connection.transaction()?;
+        ContractCodeTable::exists(&transaction, &classes)
+    })
+    .with_context(|| format!("Query storage for existance of classes {:?}", classes))?;
+    anyhow::ensure!(
+        exists.len() == classes.len(),
+        "Length mismatch when querying for class existance. Expected {} but got {}.",
+        classes.len(),
+        exists.len()
+    );
+    let missing = classes
+        .into_iter()
+        .zip(exists.into_iter())
+        .filter_map(|(class, exist)| (!exist).then(|| class));
+
+    // For each missing, download, verify and insert definition.
+    for class_hash in missing {
+        let definition = sequencer
+            .class_by_hash(class_hash)
+            .await
+            .with_context(|| format!("Downloading class {}", class_hash.0))?;
+
+        // Parse the contract definition for ABI, code and calculate the class hash. This can
+        // be expensive, so perform in a blocking task.
+        let extract = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
```

Comment:
Just curious is there any reason for `spawn_blocking` here while using `block_in_place` above?

Reply:
I think we are supposed to prefer `spawn_blocking` but I am forced to use `block_in_place` as the database transaction is `!send` and `!sync`. I could try `Arc<Transaction>` or something 🤔